### DISCUSSION TITLE
Error in Meshing on Ubuntu 22.04

### DISCUSSION THREAD

**Discussion by danishTUE** [neperfepx/neper#493] [Q&A] [answered]
Hi,

I have installed Neper on a new machine with Ubuntu 22.04. I have a script that works completely fine on a different machine with Ubuntu 18.04 and 20.04 but on Ubntu 22.04 it gives the following error when the meshing module goes from 2D meshing to 3d meshing.

Info   :   - Fixing 2D-mesh pinches...
Info   :   - 3d meshing... Error  : Wrong mesh dimension: -1!
Error  : Wrong mesh dimension: -1!
Aborted (core dumped)

I am not sure if its because of incorrect installation. However, I had installed almost all the required dependencies via the following commands:

sudo apt-get install gsl-bin
sudo apt install libomp-dev
sudo apt install libgsl-dev
sudo apt-get install libpthread-stubs0-dev
sudo apt install gmsh
sudo apt-install povray
sudo apt-get install povray
sudo apt-get install libnlopt0
git clone https://github.com/neperfepx/neper.git
cd neper/
cd src/
mkdir build
cd build
cmake ..
make
sudo make install
make test

Tests resulted in some failed tests as well. Most of the failed tests were from M and V module. 

Can you please help me in figuring out what may be missing? 

Many thanks

**Comment by danishTUE** [comment]
I am attaching the log file for the test. The log file shows that most of these test resulted in the same error
[LastTest.log](https://github.com/neperfepx/neper/files/8892669/LastTest.log)
.

**Comment by xshang93** [comment]
Did you test if the dependant, in particular gmsh, is working properly?

**Reply by danishTUE** [reply]
How can I test that? When I type gmsh in the command window, the GUI of gmsh gets opened up. So i presume its working fine?

**Comment by rquey** [comment, marked as answer]
Try to run the following:
```
export OMP_NUM_THREADS=1
gmsh --version
neper -T -n 2
neper -M n2-id1.tess -tmp .
```
After Neper crashes, there should be a file named `tmp...-1.geo` and a file named  `tmp...-1.msh` in the working directory. Please provide your terminal output and the tmp files.

**Reply by danishTUE** [reply]
Dear Romain,

Thanks for your reply.

The terminal output is as follows:-

```
4.8.4

========================    N   e   p   e   r    =======================
Info   : A software package for polycrystal generation and meshing.
Info   : Version 4.4.2-21
Info   : Built with: gsl|muparser|opengjk|openmp|nlopt|libscotch (full)
Info   : Running on 1 threads.
Info   : <https://neper.info>
Info   : Copyright (C) 2003-2022, and GNU GPL'd, by Romain Quey.
Info   : No initialization file found (`/home/dkhan/.neperrc').
Info   : ---------------------------------------------------------------
Info   : MODULE  -T loaded with arguments:
Info   : [ini file] (none)
Info   : [com line] -n 2
Info   : ---------------------------------------------------------------
Info   : Reading input data...
Info   : Creating domain...
Info   : Creating tessellation...
Info   :   - Setting seeds... 
Info   :   - Generating crystal orientations...
Info   :   - Running tessellation...
Info   : Writing results...
Info   :     [o] Writing file `n2-id1.tess'...
Info   :     [o] Wrote file `n2-id1.tess'.
Info   : Elapsed time: 0.011 secs.
========================================================================


========================    N   e   p   e   r    =======================
Info   : A software package for polycrystal generation and meshing.
Info   : Version 4.4.2-21
Info   : Built with: gsl|muparser|opengjk|openmp|nlopt|libscotch (full)
Info   : Running on 1 threads.
Info   : <https://neper.info>
Info   : Copyright (C) 2003-2022, and GNU GPL'd, by Romain Quey.
Info   : No initialization file found (`/home/dkhan/.neperrc').
Info   : ---------------------------------------------------------------
Info   : MODULE  -M loaded with arguments:
Info   : [ini file] (none)
Info   : [com line] n2-id1.tess -tmp .
Info   : ---------------------------------------------------------------
Info   : Reading input data...
Info   :   - Reading arguments...
Info   : Loading input data...
Info   :   - Loading tessellation...
Info   :     [i] Parsing file `n2-id1.tess'...
Info   :     [i] Parsed file `n2-id1.tess'.
Info   : Meshing...
Info   :   - Preparing... (cl = 0.3969) 100%
Info   :   - 0d meshing... 100%
Info   :   - 1d meshing... 100%
Info   :   - 2d meshing... 100% (0.44|0.79/100%| 0%| 0%)
Info   :   - Fixing 2D-mesh pinches...
Info   :   - 3d meshing... Error  : Wrong mesh dimension: -1!
Aborted (core dumped)
```

Also attached are the generated files in zipped format

[Output-files.zip](https://github.com/neperfepx/neper/files/8941598/Output-files.zip)

**Reply by rquey** [reply]
Your Gmsh does not seem to work properly. Try to run `gmsh -3 tmp36784-0.geo` to see if it generates a 3D mesh in `tmp36784-0.msh`, but I think it won't. You will have to install Gmsh in a different way - try to use a binary from their website or (better) compile it from the sources, taking care of resolving the dependencies.

**Reply by danishTUE** [reply]
Dear Romain, Many thanks! Using the binary from the website solved the issue. Thanks!

**Reply by rquey** [reply]
Good to know. Note that Gmsh (and therefore Neper's meshing) will run significantly faster (x2) if you compile Gmsh from the sources.

**Reply by ngrilli** [reply]
Dear @rquey and @danishTUE 

I am having the same problem on a new Ubuntu 22 installation.
Could you please specify what you mean by "use the binary"?

I have downloaded the latest gmsh version from here:
https://gmsh.info/#Download
It is version 4.10.5

I only tried to alias the name "gmsh" in my bash_aliases with the corresponding executable,
so that I can see the executable "gmsh" in my terminal,
then I reinstalled neper but
that does not change the neper behavior.
I keep getting the same error.

Once I download the gmsh binaries, how do I link it to neper?

Thank you very much,
Best Regards,
Nicolò

**Reply by ngrilli** [reply]
Answering myself: I compiled gmsh from source and it works
I did:
1) download gmsh source code here: https://gmsh.info/
2) Move to project folder
3) Follow instructions in ~/projects/gmsh-4.10.5-source/README.txt
that is:
cd ~/projects/gmsh-4.10.5-source/
mkdir build
cd build
cmake ..
make
sudo make install
This will create the executable gmsh and you can launch gmsh from terminal.
Then redo the procedure explained by @danishTUE but skip "sudo apt install gmsh"
redo neper installation, now it works.
The only thing I don't understand is that by compiling gmsh from source,
when I launch gmsh from terminal, that does not open the graphical interface.
Best Regards,
Nicolò

**Reply by rquey** [reply]
> The only thing I don't understand is that by compiling gmsh from source,
> when I launch gmsh from terminal, that does not open the graphical interface.

You probably did not install FLTK (listed as dependency when you do `cmake ..` in Gmsh).

> Once I download the gmsh binaries, how do I link it to neper?

Neper uses the binary, there is no need to link (or reinstall Neper, after you install Gmsh).

**Reply by danishTUE** [reply]
Thanks Nicolo for giving out the procedure for compiling gmsh from source. I was not able to do that and hence I had used binary. I just downloaded the binary and exported its bin path in the bashrc. After that the call to gmsh was always made to the gmsh binary i downloaded which made the neper work.

**Reply by liu373994209** [reply]
Dear danish TUE, I have encountered the same problem as you. Can you provide me with detailed guidance on how to proceed

**Reply by rquey** [reply]
https://neper.info/doc/tutorials/install_ubuntu22.html#installing-gmsh
